src/sandbox.rkt

```diff
@@ -194,69 +182,63 @@
+
+  ;; optionally compute error/cost for input expression
+  (define target-alt-data
+    (cond
+      [(test-output test)
+       (define target-expr (test-output test))
+       (define target-train-errs (errors target-expr train-pcontext ctx))
+       (define target-test-errs (errors target-expr processed-test-pcontext ctx))
+       (alt-result (make-alt target-expr) target-train-errs target-test-errs)]
+      [else
+       #f]))
 
-  (define end-errs
-    (flip-lists
-     (batch-errors (map alt-expr alts) processed-test-pcontext context)))
+  ;; compute error/cost for output expression
+  (define end-exprs (map alt-expr end-alts))
+  (define end-train-errs (flip-lists (batch-errors end-exprs train-pcontext ctx)))
+  (define end-test-errs (flip-lists (batch-errors end-exprs processed-test-pcontext ctx)))
+  (define end-alts-data (map alt-result end-alts end-train-errs end-test-errs))
 
+  ;; bundle up the result
   (timeline-adjust! 'regimes 'name (test-name test))
   (timeline-adjust! 'regimes 'link ".")
 
-  (define-values (points exacts) (get-p&es train-pcontext))
-  (define-values (newpoints newexacts) (get-p&es processed-test-pcontext))
-  (test-success test
-                (bf-precision)
-                #f
-                (timeline-extract)
-                (warning-log) (make-alt (test-input test)) alts
-                (*herbie-preprocess*) points exacts
-                (errors (test-input test) train-pcontext context)
-                (errors (alt-expr (car alts)) train-pcontext context)
-                newpoints newexacts
-                (errors (test-input test) processed-test-pcontext context)
-                end-errs
-                (if (test-output test)
-                    (errors (test-output test) processed-test-pcontext context)
-                    #f)
-                (expr-cost (test-input test) output-repr)
-                (map (curryr alt-cost output-repr) alts)
-                (*all-alts*)))
-
-;; Ugly, but struct-copy doesn't do the right thing with inheritance
-(define (add-time result time)
-  (match-define (test-success test bits _time timeline warnings
-                              start-alt end-alts preprocess points exacts
-                              start-est-error end-est-error newpoints newexacts
-                              start-error end-errors target-error
-                              start-cost end-costs all-alts) result)
-  (test-success test bits time timeline warnings
-                start-alt end-alts preprocess points exacts
-                start-est-error end-est-error newpoints newexacts
-                start-error end-errors target-error
-                start-cost end-costs all-alts))
+  (define warnings (warning-log))
+  (define timeline (timeline-extract))
+  (define pctxs (list train-pcontext processed-test-pcontext))
+  (test-result test 'success #f timeline warnings #f preprocess
+               pctxs start-alt-data target-alt-data end-alts-data))
```

Comment:
Ok this is interesting. Why use the `processed-test-pcontext` here? I see, actually. Oof, that's annoying. Preprocessing is a mess right now. By rights, it should be part of the program, a `(herbie-preprocess preprocessing ... body)` or something. Until then, I'm willing to leave it like this, even though this is pretty ugly stuff.